### ISSUE TITLE
Loopback 4 authentication Argument of type 'typeof CustomUserService' is not assignable to parameter of type 'Constructor<UserService<User, Credentials>>'.

### ISSUE THREAD

**Issue by pratikjaiswal15** [strongloop/loopback-next#5541]
I have followed this [link](https://github.com/strongloop/loopback-next/tree/master/extensions/authentication-jwt#customizing-user) for creating custom service for authentication in loopback 4. I successfully created the service but while binding this service in application.ts I am getting the following error. 

> 
> Argument of type 'typeof CustomUserService' is not assignable to parameter of type 'Constructor>'.
> Construct signature return types 'CustomUserService' and 'UserService' are incompatible. The types returned by 'verifyCredentials(...)' are incompatible between these types. Type 'Promise' is not assignable to type 'Promise'. Property 'userCredentials' is missing in type 'import("d:/ionic-pr/loopback-projects/custom/src/models/user.model").User' but required in type 'import("d:/ionic-pr/loopback-projects/custom/node_modules/@loopback/authentication-jwt/dist/models/user.model").User'

**GIT REPOSITORY** - https://github.com/pratikjaiswal15/loopbck4uth
I have two models user and userCred with has one realtionship. UserCred model contains password of the user.  The relation name is userCred.

here is the code

thank you in advance

application.ts

```
import {UserCredRepository, UserRepository} from './repositories';
import {CustomUserService} from './services/custom-user.service';
import {UserServiceBindings} from '@loopback/authentication-jwt';

// Bind user service inside constructor
    this.bind(UserServiceBindings.USER_SERVICE).toClass(CustomUserService),// error on this line

      // Bind user and credentials repository
      this.bind(UserServiceBindings.USER_REPOSITORY).toClass(
        UserRepository,
      ),

      this.bind(UserServiceBindings.USER_CREDENTIALS_REPOSITORY).toClass(
        UserCredRepository,
      )
```

custom-service.ts

```
import {UserService} from '@loopback/authentication';
import {repository} from '@loopback/repository';
import {HttpErrors} from '@loopback/rest';
import {securityId, UserProfile} from '@loopback/security';
import {compare} from 'bcryptjs';
// User --> MyUser
import {User} from '../models/user.model';
// UserRepository --> MyUserRepository
import {Credentials, UserRepository} from '../repositories/user.repository';


export class CustomUserService implements UserService<User, Credentials> {
  constructor(
    // UserRepository --> MyUserRepository
    @repository(UserRepository) public userRepository: UserRepository,
  ) {}

  // User --> MyUser
  async verifyCredentials(credentials: Credentials): Promise<User> {
    const invalidCredentialsError = 'Invalid email or password.';

    const foundUser = await this.userRepository.findOne({
      where: {email: credentials.email},
    });
    if (!foundUser) {
      throw new HttpErrors.Unauthorized(invalidCredentialsError);
    }

    const credentialsFound = await this.userRepository.findCredentials(
      foundUser.id,
    );
    if (!credentialsFound) {
      throw new HttpErrors.Unauthorized(invalidCredentialsError);
    }

    const passwordMatched = await compare(
      credentials.password,
      credentialsFound.password,
    );

    if (!passwordMatched) {
      throw new HttpErrors.Unauthorized(invalidCredentialsError);
    }

    return foundUser;
  }

  // User --> MyUser
  convertToUserProfile(user: User): UserProfile {

    let address = ''
    if (user.address) {
      address = user.address
    }

    return {
      [securityId]: user.id!.toString(),
      name: user.name,
      id: user.id,
      email: user.email,
      address: address // this are all fields of user model. Address is not required
    };
  }
}
```

user.repositor.ts

```
import {Getter, inject} from '@loopback/core';
import {DefaultCrudRepository, HasOneRepositoryFactory, repository} from '@loopback/repository';
import {MygodDataSource} from '../datasources';
import {User, UserCred, UserRelations} from '../models';
import {UserCredRepository} from './user-cred.repository';

export type Credentials = {
  email: string;
  password: string;
};



export class UserRepository extends DefaultCrudRepository<
  User,
  typeof User.prototype.id,
  UserRelations
  > {

  public readonly userCred: HasOneRepositoryFactory<UserCred, typeof User.prototype.id>;

  constructor(
    @inject('datasources.mygod') dataSource: MygodDataSource, @repository.getter('UserCredRepository') protected userCredRepositoryGetter: Getter<UserCredRepository>,
  ) {
    super(User, dataSource);
    this.userCred = this.createHasOneRepositoryFactoryFor('userCred', userCredRepositoryGetter);
    this.registerInclusionResolver('userCred', this.userCred.inclusionResolver);
  }


  async findCredentials(
    userId: typeof User.prototype.id,
  ): Promise<UserCred | undefined> {
    try {
      return await this.userCred(userId).get();
    } catch (err) {
      if (err.code === 'ENTITY_NOT_FOUND') {
        return undefined;
      }
      throw err;
    }
  }
}
```

**Comment by pratikjaiswal15** [comment]
Please find git repository https://github.com/pratikjaiswal15/loopbck4uth

**Comment by madaky** [comment]
If you look at the error you will get clear vision of what you are missing: So lets proceed line by line.

Construct signature return types 'CustomUserService' and 'UserService' are incompatible.
Explanation : This clears you that the services that you are trying to bind has signature difference which exactly means the fingerprint are mismatch..

The types returned by 'verifyCredentials(...)' are incompatible between these types
Explanation: This takes you one step closer to problem definition. It says type return by verifyCredentials(..) are incompatible. As you can see you are returning the User in the listed method. So it gives you clear view about you are missing the typed properties which are mandatory to use in your User model.

type 'Promise' is not assignable to type 'Promise'. Property 'userCredentials' is missing in type 'import("d:/ionic-pr/loopback-projects/custom/src/models/user.model").User' but required in type 'import("d:/ionic-pr/loopback-projects/custom/node_modules/@loopback/authentication-jwt/dist/models/user.model").User'
Explanation : This make you clear that which property signature you are missing see I make it in bold. So it is clear that you are missing the 'userCredentials" in you user.model.ts so

@hasOne(() => UserCred, {keyTo: 'user_id'})
  userCredentials: UserCred;
So Basically its the summary of the error you are facing now.

But changing this will not solve your problem. You will discover new errors. I can tell you right away now, but would not do so. Because I want you to explore the mistakes that you are doing and resolve it by your own. One thing that I observed

@property({
    type: 'number',
    id: true,
    generated: true,
  })
  id?: number;
id is something which should be mandatory in your model, because there is no other model as you don't have any other key in the model. So why there is a "?". Its a bad software practice. Well I think this much explanation is enough for you to proceed, and make your code working.

!!Happy Coding!!

**Comment by pratikjaiswal15** [comment]
Thank a lot @madaky. I appreciate it. I had to change has one property name from `userCred `to  `userCredentials` to get rid of that issue. But as you stated I have got many errors and I tried to correct them. Correct me if I am wrong

1. Initially I got error with id. If I don't use ! while returning from custom-service then I was getting `object is possibly undefined` error. So I have to remove ! operator from id property like below

```
@property({
    type: 'number',
    id: true,
    generated: true,
  })
  id: number; //notice no !
```
2. Then I got compatibility error with `user_id` in the user-credential model. I have to change the property name from `user_id `to `userId ` to get rid of that issue. (I saw it in user-credential model from the authentication-jwt package)

3. Now I am getting a strange compatibility issue with userCredential. And actually I can't get why this error is occurring. Can you please help? 
Thank you so much, 
```
Argument of type 'typeof CustomUserService' is not assignable to parameter of type 'Constructor<UserService<User, Credentials>>'.
  Construct signature return types 'CustomUserService' and 'UserService<User, Credentials>' are incompatible.
    The types returned by 'verifyCredentials(...)' are incompatible between these types.
      Type 'Promise<import("d:/ionic-pr/loopback-projects/custom/src/models/user.model").User>' is not assignable to type 'Promise<import("d:/ionic-pr/loopback-projects/custom/node_modules/@loopback/authentication-jwt/dist/models/user.model").User>'.
        Type 'import("d:/ionic-pr/loopback-projects/custom/src/models/user.model").User' is not assignable to type 'import("d:/ionic-pr/loopback-projects/custom/node_modules/@loopback/authentication-jwt/dist/models/user.model").User'.
          The types of 'userCredentials.id' are incompatible between these types.
            Type 'number | undefined' is not assignable to type 'number'.
              Type 'undefined' is not assignable to type 'number'.
```

**Comment by madaky** [comment]
@pratikjaiswal15 : Notice this
 ** The types of 'userCredentials.id' are incompatible between these types.
            Type 'number | undefined' is not assignable to type 'number'.**
you will find answer. :)

**Comment by pratikjaiswal15** [comment]
Thank you so much. I resolved everything.

**Comment by llhilton** [comment]
Any chance you can tell what the issue is? I got the same error. It seems to be upset that the credentials I need in my custom service don't match those in the default user service. But the whole point of a custom service should surely be to have my own custom credentials, right? Maybe I'm missing something else (like a credential repository, but I have no idea why I'd need one), but it's completely unclear to me what. (This is my first node framework, so it could be something really obvious to others what is opaque to me.) 

Full error:
src/application.ts:53:57 - error TS2345: Argument of type 'typeof AffiliateService' is not assignable to parameter of type 'Constructor<UserService<User, Credentials>>'.
  Construct signature return types 'AffiliateService' and 'UserService<User, Credentials>' are incompatible.
    The types of 'verifyCredentials' are incompatible between these types.
      Type '(credentials: Credentials) => Promise<Affiliate>' is not assignable to type '(credentials: Credentials) => Promise<User>'.
        Types of parameters 'credentials' and 'credentials' are incompatible.
          Property 'ORG_UNIT_ID' is missing in type 'import("/home/lisa/sites/member.application.backend/node_modules/@loopback/authentication-jwt/dist/services/user.service").Credentials' but required in type 'import("/home/lisa/sites/member.application.backend/src/services/affiliate.service").Credentials'.

53     this.bind(UserServiceBindings.USER_SERVICE).toClass(AffiliateService),
                                                           ~~~~~~~~~~~~~~~~

  src/services/affiliate.service.ts:9:3
    9   ORG_UNIT_ID: string,
        ~~~~~~~~~~~
    'ORG_UNIT_ID' is declared here.


affiliate.service.ts: 
`import {UserService} from '@loopback/authentication';
import {repository} from '@loopback/repository';
import {HttpErrors} from '@loopback/rest';
import {securityId, UserProfile} from '@loopback/security';
import {Affiliate} from '../models';
import {AffiliateRepository} from '../repositories';

export type Credentials = {
  ORG_UNIT_ID: string,
};

export class AffiliateService implements UserService<Affiliate, Credentials> {
  constructor(
    // UserRepository --> MyUserRepository
    @repository(AffiliateRepository) public affiliateRepository: AffiliateRepository,
  ) {}

  // User --> MyUser
  async verifyCredentials(credentials: Credentials): Promise<Affiliate> {
    const invalidCredentialsError = 'Invalid affiliate identification.';

    const foundAffiliate = await this.affiliateRepository.findOne({
      where: {ORG_UNIT_ID: credentials.ORG_UNIT_ID},
    });
    if (!foundAffiliate) {
      throw new HttpErrors.Unauthorized(invalidCredentialsError);
    }
    return foundAffiliate;
  }

  // User --> MyUser
  convertToUserProfile(affiliate: Affiliate): UserProfile {
    return {
      [securityId]: affiliate.ORG_UNIT_ID,
      ORG_UNIT_ID: affiliate.ORG_UNIT_ID,
      ORG_UNIT_NAME: affiliate.ORG_UNIT_NAME
    };
  }
}
`

JWT bindings in application.ts constructor:
`    // jwt bindings
    this.component(AuthenticationComponent);
    // Mount jwt component
    this.component(JWTAuthenticationComponent);
    // Bind datasource
    this.dataSource(MemberApplicationDataSource, UserServiceBindings.DATASOURCE_NAME);
    this.bind(UserServiceBindings.USER_SERVICE).toClass(AffiliateService),
      // // Bind user and credentials repository
      this.bind(UserServiceBindings.USER_REPOSITORY).toClass(AffiliateRepository);
`

**Comment by pratikjaiswal15** [comment]
@llhilton The thing is the fields and data types in your model need to be same as user model in `@authentication-jwt` package.
`ORG_UNIT_ID` should be present in user model model of authentication-jwt package with same data type

**Comment by magixus** [comment]
yeah putting same Id type solved the issue, for me thanks yall
